ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java

```diff
@@ -125,234 +122,281 @@
-                    //set custom cell renderer
-                    comboBox.setCellFactory((ListView<Optional<DataSource>> param) -> new DataSourceCell(dataSourcesTooManyFiles, controller.getAllDataSourcesDrawableDBStatus()));
-                    comboBox.setButtonCell(new DataSourceCell(dataSourcesTooManyFiles, controller.getAllDataSourcesDrawableDBStatus()));
-
-                    DataSource dataSource = datasourceDialog.showAndWait().orElse(Optional.empty()).orElse(null);
-                    try {
-
-                        igTopComponent.showDataSource(dataSource, dataSourcesTooManyFiles);
-                    } catch (TskCoreException ex) {
-                        if (dataSource != null) {
-                            logger.log(Level.SEVERE, "Error showing data source " + dataSource.getName() + ":" + dataSource.getId() + " in Image Gallery", ex);
-                        } else {
-                            logger.log(Level.SEVERE, "Error showing all data sources in Image Gallery.", ex);
-                        }
-                    }
-                });
-
-                return null;
-            }
-        }).start();
     }
 
-    synchronized private void showDataSource(DataSource datasource, Map<DataSource, Boolean> dataSourcesTooManyFiles) throws TskCoreException {
-        if (dataSourcesTooManyFiles.get(datasource)) {
+    /**
+     * Configures the groups manager for the selected data source(s) and opens
+     * this top component's window.
+     *
+     * @param selectedDataSource      The data source selected, null if all data
+     *                                sources are selected.
+     * @param dataSourcesTooManyFiles A map of data sources to flags indicating
+     *                                whether or not the data source has to many
+     *                                files to actually be displayed.
+     */
+    private void openWithSelectedDataSources(DataSource selectedDataSource, Map<DataSource, Boolean> dataSourcesTooManyFiles) {
+        if (dataSourcesTooManyFiles.get(selectedDataSource)) {
             Platform.runLater(ImageGalleryTopComponent::showTooManyFiles);
-            return;
-        }
-        // Display the UI so that they can see the progress screen
-        SwingUtilities.invokeLater(() -> showTopComponent(this));
-        GroupManager groupManager = controller.getGroupManager();
-        synchronized (groupManager) {
-            groupManager.regroup(datasource, groupManager.getGroupBy(), groupManager.getSortBy(), groupManager.getSortOrder(), true);
+        } else {
+            /*
+             * Open the top component's window before configuring the groups
+             * manager so that the spinner(s) that take the place of a wait will
+             * be displayed if the operations takes awhile.
+             */
+            // RJCTODO: Is this really necessary?
```

Comment:
the call to showTopComponent()?  I think it is needed, or something similar, to bring the TopComponent to the front if it is already open and the user tries to open it again.

Reply:
Okay.